### PR TITLE
Move badge_type logic into server

### PR DESCRIPTION
**Pull request by kitsuta**
This is the first step to refactor the badge_type selector. The basic logic of displaying and configuring badges has been moved to the Session class. Dummy displays have also been added to the correct templates for testing.

Todo before merging:
* figure out how to more dynamically define/configure the donation-level-based badge types
* determine whether an attendee has already selected an option (or donated) and send along the badge type's starting status (disabled, enabled, selected) accordingly

I'd like @EliAndrewC to review the current function, as it can probably be improved. @wesleycho can fork this branch and get started on the front-end if he's got the time/inclination.

### CODE REVIEW

File: uber/models.py [hunk 1]
```diff
@@ -470,6 +470,73 @@ def jobs_for_signups(self):
...
+            There are several contexts where we want to display different badge types to select:
+            - A new attendee registering for the event
+            - An attendee claiming a blank badge in an existing group
+            - An existing attendee editing their registration
+
+            This function uses the parameters in the session to figure out the context and return the correct
+            badge types to display for an attendee.
+            """
+            params = cherrypy.lib.httputil.parse_query_string(cherrypy.request.query_string)
+            group_id = params['group_id'] if 'group_id' in params else None
```
Comment (kitsuta) [inline uber/models.py]: These can probably be combined so there's fewer interim steps - I just wanted to push out a working version of the code, first. :)

**Comment by EliAndrewC** [comment]
What's the motivation behind doing this in the session class, or more generally in Python code in general?  I get that we sometimes want to show different options, like we want to show the Group option on the main prereg form but not in other places.  But why not put that logic in the template?  If we're literally generating HTML code like ``<p class="list-group-item-text">...`` then the SQLAlchemy Session mixin class seems like a less intuitive place for the logic than the template (or a custom tag or whatever).

(Minor related note: One feature that's currently implemented but not used is that an entry in ``BADGE_TYPES`` can define a callback function which is invoked whenever that option is selected.  Maybe we can have that functionality go away, since it's not currently used.  The idea is that if MAGStock wanted to make "Food" a top-level choice, then clicking that option would cause the "Food" checkbox to be checked, etc.  But since we didn't end up doing that, maybe it's not a huge deal.  I think Anthrocon might end up finding it useful, but I don't remember offhand whether that would be helpful for them; MAGFest events are probably fine without it for the time being.)

Regardless, I'm mostly indifferent to whether this lives on the frontend or backend, since the proof is in the pudding, as it were.  If it ends up being as simple or simpler for plugins to add/remove/modify entries in the badge type tables, then it'll be fine.  If not then we'll probably need to move this logic to the template, since ``BADGE_TYPES`` has the advantage of being easily modifiable in ``regextras.html``.

**Comment by kitsuta** [comment]
I don't know about it being "easily modifiable" in the templates since I have been completely unsuccessful in actually modifying it in the Anthrocon plugin. :P For backend developers like me, it kinda sucks to have to monkey around this much with jQuery just to change what gets shown. And for frontend developers, ... well they're probably fine, really.

The short answer is that generating these options serverside gives us more flexibility in defining what they are, what they display, and when they display what. Right now this is largely just a copy of the options/logic in the form.html, so it needs some work - in fact, that HTML probably shouldn't be included in the dict. But I wanted to take the heavy lifting logic (particularly in deciding what to display for the badge name and whether or not to display certain options) out of the templates. A dict also separates the badge choices from the display logic, so if an event wanted to it could more easily reformat the look of the badge options (e.g., it could automatically show up as a dropdown if there are tons of "badge options" configured).

Making it a custom tag could be the best choice here, since badge options are a weird intersection of view-logic and server-logic and configuration.

**Comment by wesleycho** [comment]
The data itself should not live in the view - that is hiding a data model & business logic relating to that model in the view layer, which mixes concerns.

Having the model in the backend allows for more flexibility in general, including if there is ever a desire to have the backend become an api platform. The template generation also becomes much more simplified. In fact, it becomes more possible to even write a smarter template and kill this JavaScript usage.

I'm not sure if it belongs in a session mixin though - maybe a helper service for preregistration where it uses the session information to determine what badges to list?

**Comment by kitsuta** [comment]
On a personal note, you, uh, kinda ripped out a bunch of important functionality I had programmed in the badge_choice template, which leads me to believe the function wasn't very readable or clear about what it was doing. Besides readability, the attachment of decision logic to view logic meant that a view refactor destroyed existing functionality. Which is... bad.

To clarify, the things we lost were:
* the ability to dynamically display the person's actual or expected badge type - Staff, Dealer, etc. That's now back in this function. 
* the ability to "lock down" badges you've upgraded past (so if you're a Season Supporter, you can't ever select the Supporter badge, etc). I still need to figure out how to re-add it.

**Comment by EliAndrewC** [comment]
> The data itself should not live in the view - that is hiding a data model & business logic relating to that model in the view layer, which mixes concerns.

Ehh, I get what you're saying, but in practice the lines are often much fuzzier than that.  If I have a dropdown, and changing the value of that dropdown causes different fields to be shown or hidden, then where does that logic "belong"?  Almost everyone I know would use Javascript on the frontend for that.

But what if the values of that dropdown are determined by a combination of how the server is configured and what's in the database?  Those actual options should almost certainly be generated on the backend.

This sort of thing often puts us in situations where you've got frontend code which needs to be aware of certain aspects of the backend (hopefully in a limited way).  Or vice versa, possibly.

In general, "I have a form, and clicking on different things needs to dynamically change the form" seems like a problem which can be reasonably solved by "view logic".  Or not.  I'm not ideological about it, so in my mind whatever produces the simplest and most easily extensible code wins.  I strongly suspect in this case that just letting plugins modify a global Javascript object on those pages is going to win in terms of simplicity and extensibility, but I'll be happy to be shown otherwise :)

**Comment by kitsuta** [comment]
My ideal for extensibility is to make it so that anything that's not "an individual" or "a group" is a config dict that automatically gets appended to the base dict (which, by default, contains "Attendee" and "Group Leader"). So plugins can throw dicts at the system when they want different badge types.

Possibly we want to do this through flatfile config, so you have something like:
```
[[donation_badge_options]]
[supporter_option]
name = "Supporter"
description = "words"
extra = SUPPORTER_LEVEL
dependency = SUPPORTER_LEVEL
```
and we do some minor logic in the Session class (or a custom tag, etc) where `extra` gets appended to a base badge price and we automatically check for `extra` existing in PREREG_DONATION_TIERS and we check if the dependency is `_AVAILABLE`

I have probably just way overengineered that but yeah.

**Comment by kitsuta** [comment]
>In general, "I have a form, and clicking on different things needs to dynamically change the form" seems like a problem which can be reasonably solved by "view logic".

The actual problem here is "I have a form, and depending on the attendee's data, the context the form is being accessed in, and various config options, that form should display different things."

The badge choice function seems really simple until you start trying to implement it for pages that aren't the pre-reg form. This is actually exactly what happened with https://github.com/magfest/ubersystem/blob/ac-merge/uber/templates/preregistration/badge_choice.html - I decided to implement badge choices on the front end, and by the time I realized that was probably a bad idea, I had a giant mess of JavaScript on my hands. So this is essentially a refactor of that.

I created a PR while it was incomplete because I wanted a sanity check on the direction the refactor was going in - you're both right in that the Session class isn't exactly an obvious place for it. I'll likely look into making it a custom tag... later, when I have brain power again. :)

**Comment by EliAndrewC** [comment]
Gotcha.  Lemme know if you get stuck - I definitely understand (now) that my refactor eliminated some things it turns out we want (such as being able to show that header on other pages, which I didn't think was a thing but makes sense as something we actually want to do).  So I won't mind if I end up needing to pitch in to clean up my own mess :)

**Comment by kitsuta** [comment]
OK, I've thought about it a bit more and here's my proposal:

I'd like to add `button_opts` to custom_tags.py. It should be able to handle:
- a dict of dicts
- a single dict (assumed to contain just names/prices, much like donation tiers are now)
- a list (same as single dict)

It handles the front-end stuff in much the same way that the current `options` custom tag does. Optionally, I'd like to adjust the current `options` tag to also be able to take the same forms of data. Also optionally, it'd be nice to change its name to `dropdown_opts`.

The functionality that currently exists in the `Session` class will be moved to config.py where it belongs. It'll be a global property that bases the data on flatfile config.

The functionality I haven't implemented yet, that of selecting the right badge type and disabling badge types you can't downgrade to, will be added in the custom tags - definitely to `button_opts` and hopefully also to `dropdown_opts`.

I know it's a lot of engineering, but I think this does what we want, puts configuration where it belongs, and takes the logic out of the templates themselves.

File: uber/models.py [hunk 2]
```diff
@@ -1135,10 +1135,6 @@ def shirt_size_marked(self):
         return self.shirt not in [c.NO_SHIRT, c.SIZE_UNKNOWN]
 
     @property
-    def is_group_leader(self):
```
Comment (kitsuta) [inline uber/models.py]: This was an unrelated fix. The exact same function is also just above `shirt_size_marked`, so this was a duplicate.

File: uber/config.py [hunk 3]
```diff
@@ -263,6 +322,12 @@ def _is_intstr(s):
 for _badge_type, _range in _config['badge_ranges'].items():
     c.BADGE_RANGES[getattr(c, _badge_type.upper())] = _range
 
+_make_enum('badge_display', OrderedDict([(name, section['name']) for name, section in _config['badge_display'].items()]))
```
Comment (kitsuta) [inline uber/config.py]: I pretty much just copied the code for age groups to make this. We could use better support for defining dicts within dicts inside the configuration.

File: uber/config.py [hunk 4]
```diff
@@ -72,6 +72,65 @@ def PREREG_BADGE_TYPES(self):
...
+
+            if c.GROUPS_ENABLED and not attendee and not group:
+                badge_types['group'] = {
+                    'title': 'Group Leader',
+                    'description': group_description,
+                    'extra': 0
+                }
+
+            # The rest of the badges are added in via config
+            for name, option in c.BADGE_DISPLAY_CONFIGS.items():
```
Comment (kitsuta) [inline uber/config.py]: This takes a special set of options in development.ini and appends them to the badge_types dict. For example, using this config:

```
[badge_display]
[[supporter]]
name = "Supporter"
description = "Donate extra and get more swag with your membership."
extra = 50

[[season]]
name = "Season Supporter"
description = "Fill this in later."
extra = 100
```

will create two more options.

I'd like to be able to set `extra` to an integer enum, but trying to do so failed, and fixing that seems excessive. You'll just have to make sure they match your donation tiers.

File: uber/config.py [hunk 5]
```diff
@@ -72,6 +72,65 @@ def PREREG_BADGE_TYPES(self):
...
+                if int(option['extra']) in c.PREREG_DONATION_TIERS:
```
Comment (kitsuta) [inline uber/config.py]: This function is missing the ability to check the availability of certain levels. I haven't yet decided the best way to support configuring this, since the _AVAILABLE properties themselves are dynamic (and, specifically, aren't guaranteed to exist).

Comment (kitsuta) [inline uber/config.py]: Oh wait, the availability checks are inside PREREG_DONATION_OPTS, so this inherits that. Nevermind then, it works perfectly.

**Comment by kitsuta** [comment]
OK, the backend is completed.

Instead of a custom tag for displaying these buttons I opted to more directly interface with the Javascript that was already there. I also moved said JavaScript into preregbase.html.

There are a few front-end bugs:
1. Clicking the "group" box sets the Attendee box as being active. The correct forms load, however.
2. The page does not seem to pre-activate the right button when you have `amount_extra` set. In other words, `var makeBadgeMatchExtra` doesn't seem to work properly on anything except the pre-reg page.
3. I haven't yet figured out how to specify an order for the badge types.

Once we get those fixed, this should be good to go.

File: uber/templates/preregistration/form.html [hunk 6]
```diff
@@ -129,7 +55,7 @@
 
 <form method="post" action="form" class="form-horizontal" role="form">
 
-<input type="hidden" name="badge_type" value="{{ attendee.badge_type }}" />
+<input type="hidden" id="badge_type" name="badge_type" value="{{ attendee.badge_type }}" />
```
Comment (kitsuta) [inline uber/templates/preregistration/form.html]: Oops, I need to undo this change.

File: uber/config.py [hunk 7]
```diff
@@ -72,6 +72,72 @@ def PREREG_BADGE_TYPES(self):
...
+        There are several contexts where we want to display different badge types to select:
+        - A new attendee registering for the event
+        - An attendee claiming a blank badge in an existing group
+        - An existing attendee editing their registration
+
+        This property uses the parameters in the URL to figure out the context and return the correct
+        badge types to display for an attendee.
+        """
+        with sa.Session() as session:
+            params = cherrypy.lib.httputil.parse_query_string(cherrypy.request.query_string)
```
Comment (EliAndrewC) [inline uber/config.py]: Since this is effectively a method which responds to parameters, it would probably be more idiomatic to make this a method instead of a property and explicitly pass those as function parameters.  That would also make it possible to call this method from somewhere other than a page handler.

Comment (kitsuta) [inline uber/config.py]: But doesn't that mean you specifically _can't_ use it in the Django template? I was hoping to avoid having to add multiple instances where this function is called.

File: uber/config.py [hunk 8]
```diff
@@ -72,6 +72,72 @@ def PREREG_BADGE_TYPES(self):
...
+            donation_prepend = '' if base_badge_name == c.BADGES.get(c.ATTENDEE_BADGE) else base_badge_name + ' / '
+            badge_cost = attendee.badge_cost if attendee else c.BADGE_PRICE
+
+            # The base badge and the group badge are special, so they're given default values and added manually
+            base_description = 'Allows access to the convention for its duration.'
+
+            if c.BEFORE_GROUP_PREREG_TAKEDOWN:
+                group_description = 'Register a group of ' + str(c.MIN_GROUP_SIZE) + ' or more and save $' + \
+                                    str(c.GROUP_DISCOUNT) + ' per badge.'
+            else:
```
Comment (EliAndrewC) [inline uber/config.py]: Fun fact; Python concatenates strings by default, so in cases where you want to split a string into parts and put each part on a line like this, you don't even need the plus and can literally just say

```python
foo = ('Hello '
       'World')
```
or if you prefer
```python
foo = 'Hello ' \
      'World'
```

I'm personally more of a "just put it all on one long line" kind of guy, but that's just a style preference and our PEP8 checker has no preference :)

File: uber/config.py (hunk 7 above)
Comment (EliAndrewC) [inline uber/config.py]: Yeah, if you wanted to have arguments in a Django template you'd need to do that via a custom tag.  The more idiomatic approach would probably be to write this as a method and then have a custom tag which is just a thin wrapper around this function.  For now we should probably just get something working and then look at it once we have it in a good state.  (Side note: if I was picking a template engine today I'd pick Jinja templates, which are basically "We've copied Django templates but you can actually call methods with parameters and such.")

Comment (kitsuta) [inline uber/config.py]: Yeah, putting it in a custom tag was my first approach, but that quickly started taking too long to implement for what's really a fairly minor feature.

Jinja sounds nice. :( If we can't port this whole thing over to Django eventually, maybe we'll just refactor that in.

**Comment by thaeli** [comment]
Closing this PR so we can close out the mag-unstable branch. Please resubmit pull of this feature branch against master.
